### ISSUE TITLE
Reflecting current set of CSS selectors : context.selectors not being carried over in eval contexts

### ISSUE THREAD

**Issue by rjgotten** [less/less.js#2619]
In working on some complex framework components written in Less, I have a few times already come across situations where it would have been beneficial to be able to reflect the current set of CSS selectors into a variable for processing.

I've succesfully built a custom function that does just this, but it runs afoul of an issue deeper in the Less compiler.
### My usecase

My current case and the one I'll treat here as an example for the reflection feature and how it leads up to the problem with the compiler, is a stylized button that contains both a left/right positioned font-icon and a text span that has horizontal ellipsis applied for overflow and can have its own text left/right/center aligned within the remaining space on the button. E.g.

``` html
<button type="button" class="imp-button" data-iconpos="left">
  <i class="imp-icon" data-icon="check"></i>
  <span class="imp-button-text">OK</span>
</button>
```

Currently, I have a mixin which renders the base of the button's CSS; handling things like stripping default button styles, setting proper padding, line-height, etc. and handling the quirks of `::moz-focus-inner` on older Firefox builds.

As part of this mixin, I have to set both CSS for the button itself and for the text span inside it and both class names have a shared root in the name. (OOCSS principles. This pattern is also re-used across multiple of our sites.)

No problem right?

``` less
.lib-button(@height; @hor-padding) {
   // Set some properties for the main button here

   & > &-text {
     // And for the text span here.
   }
}
```

Call via:

``` less
.imp-button {
  .lib-button(40px, 10px);
}
```

And done:

``` css
.imp-button { }
.imp-button > .imp-button-text {}
```
#### Where my usecase breaks

This is a nice pattern, but it still breaks when we _subclass_ the button, e.g. make a compact variation:

``` less
.imp-button--compact {
  .lib-button(20px, .5px);
  font-size : 10px;
  // etc.
}
```

``` css
.imp-button--compact { }
.imp-button--compact > .imp-button--compact-text {} /* <-- Uh oh! */
```

It should generate `.imp-button--compact > .imp-button-text` instead, so that we don't have to double up on all the nested classes on subcomponents for different variations.
#### Arriving at a solution

I've come up with a few complex builder patterns to work around this, but none of them really fit the bill 100% and really, they all feel like just stacking on needless complexity, when I could solve this _tout-de-suite_ if I could have a little look at the right-hand key-selector and chop off the part after the double dash that we use in our convention and thus simply hide all the complexity from my libraries' consumers.

E.g.

``` less
.lib-button(@height; @hor-padding) {
   // Pull up the current right-hand key selectors and work with the first one.
   // (KISS: assume people aren't calling this with multiple selectors.)
   @selector : extract(key-selectors(),1);

   // Some extraction of the base and variation part of a classname. Probably regex match?
   @class-base      : ...
   @class-variation : ...

   .lib-button-text() {
     // Shared text span implementation here.
   }

   &-text when ( @class-variation = "" ) { .lib-button-text();  }
   & > @{class-base}-text  when not ( @class-variation = "" ) { .lib-button-text();  }
}


.imp-button {
  .lib-button(40px, 10px);
}

.imp-button--compact {
  .lib-button(20px, .5px);
}
```

Output:

``` css
.imp-button {}
.imp-button-text {} /* <-- Nicely shortened, removing the redundant child combinator! */

.imp-button--compact {}
.imp-button--compact > .imp-button-text {} /* <-- Perfect! */
```

I've already implemented a generic `selectors()` function to get the full set of active selectors (taking into account `&` and variable-name token replacement) and a `key-selectors()` function that does the same, but returns the key-selector (i.e. the most right-hand element) only.

It took some digging to figure out how to work with the data structures that the compiler has in place, but it was very doable as a pluggable extension function that walks back over the stack of selectors maintained in the Eval context of the function call node and assembles them into the full set of complete selectors. If that sounds as a surprise: yes they are accessible via `this.context.selectors`, as functions are executed with `this` bound to the node instance.

(Friendly Note: The fact that you can acccess the current eval context that way is also a god-sent that empowers plugin functions to fullfill several very complex scenarios. Please, please, please; don't _ever_ change that!)
### The actual problem

The problem is that this works fine for regular nested rulesets, but currently **BREAKS** when performed directly or indirectly inside a mixin or detached ruleset.

I traced down the root cause, which is the fact that when `callEval` is executed for a DetachedRuleset or MixinDefinition node, a new Eval context is created and the selectors are not part of the copy-constructed properties. Ergo, inside a mixin or ruleset call `this.context.selectors` _resets_ to the empty set.

Would there be any harm in adding `"selectors"` to the array of cloned properties for Eval contexts? (I could imagine it breaking the join-selector visitor in some way, but I'm not sure.)

**Comment by matthew-dean** [comment]
You lost me at:

> It should generate .imp-button--compact > .imp-button-text instead, so that we don't have to double up on all the nested classes on subcomponents for different variations.

I'm not sure why it should generate something other than what you've written. If you call the mixin from within ```.imp-button--compact```, then what was output seems perfectly expected.

To me, this seems like a situation of designing yourself into a corner. You designed your mixins and classes in a particular way, but then want to change the compiler from expected output?

I dunno, it just seems like the proposed solution is a product of overthinking. There are a number of other ways to get to the proposed output other than adding a new feature.

**Comment by rjgotten** [comment]
> I'm not sure why it should generate something other than what you've written.

Let me rephrase: Ideally I would like to write a mixin that generates `.imp-button--compact > .imp-button-text` instead of doubling up the `--compact` suffix and the pattern of using the `&` replacement token inside the URL selector does not agree with that.

>  There are a number of other ways to get to the proposed output other than adding a new feature.

While there are *other* ways to accomplish this than reflecting on the current selector and extracting the base of the CSS class that way, those methods are all leaky abstractions that place additional burden on the consumers of the mixin library. That's something I specifically want to try and avoid, because none of it **scales**. (Keep in mind; this presentation of my use-case is a severe reduction of what the actual thing looks like!)

As for this being a request to 'add new features': i'm not asking for new core features **at all**. I'm only asking whether the non-propagation of the selector set through mixin calls or ruleset calls is a conscious choice, or whether it is in fact an oversight or bug. (And if it's a bug: whether fixing it would break other things that are implicitly relying on the 'broken' behavior.)


> I dunno, it just seems like the proposed solution is a product of overthinking.

Reflecting or capturing the current selector has been a part of Less's main competitor Sass for a long, long time. And it's most commonly used for *exactly* the kind of pattern I'm explaining here.
It's not a product of overthinking.

**Comment by seven-phases-max** [comment]
I agree with @matthew-dean, the use-case example is too close to an XY problem. <sub>Actually for me the example looks like just a standard (one of "top 5") development pitfall: one tries to design the only function/mixin "to rule them all", and then (since this non-modular approach always fails in the end) he tries to invent some stuff to put into that "I can do everything" entity to use as conditions/exceptions for each (usually orthogonal) specific context he want that function/mixin in. (This is absolutely fine during drafting and/or when you need something fast-n-dirty to work, but this obviously not a good example to back a new language feature).</sub>

In other words, the code (using the purposed feature) in your example only tries to fight/cancel `& > &-text` of your initial mixin (i.e. instead of solving the root of the problem itself, you use the proposed feature only to work out the consequences of the initial method you used to make it in a simple case).

So not diminishing the purposed feature itself in general, I also think the example just prays for another solutions. And though I probably would sympathize some extra "selector manipulation facility" in general, if it's only about "sniffing a ruleset the mixin is invoked in and change the output depending on that" - ouch... If you need to play with selectors in a mixin pass them explicitly, e.g.:

[`1`](https://gist.github.com/seven-phases-max/b13400c1aeed8bc45448) - simple and straight-forward
[`2`](https://gist.github.com/seven-phases-max/a8b8db69e14472ab5699) - the same in reverse 
[`3`](https://gist.github.com/seven-phases-max/011d7116bb387f241a87) - same as `1` w/o "redundant `.base-button > .base-button-text`"
etc. and so on.

Obviously those are simplified examples just for illustration. Indeed, there're many ways and exact code may vastly differ depending on the actual needs. Especially considering those "stripping default button styles, the quirks of  `::moz-focus-inner`" you mentioned, which means that most likely your actual button styles differ only in a few properties and the rest is the same, so I'd expect some `extend` going inside of the mixin(s), which in its turn leads to favor other impl. methods and so on (e.g. crafting things to use the `extend` w/o bloating there may be much more important than inventing an extra code to eliminate "redundant `.base-button > .base-button-text`" etc.).

---
So in summary:
I can imagine how other new features can make my above examples more DRY or more readable (#1546, #1075 etc.).
But for this particular feature-request (in short: "allow expand/evaluate `&` in value statements"?) I'm afraid you'll need to find another use-case example (I recall there were some issues where something like this seemed to be helpful, but I can't find those right now). But so far my -1 too.

**Comment by rjgotten** [comment]
> if it's only about "sniffing a ruleset the mixin is invoked in and change the output depending on that"

It's not. Please go back and re-read. 

The fact that I change output (i.e. either render with or without child combinator) is a nice optimization extra that makes the CSS cheaper to process by a browser during the style matching phase. It's not the core motivation behind the use-case.

The core motiviation is being able to generate a child class selector based on a common prefix in the parent class selector without having to explicitly hand-type the child class selector and pass it in as a parameter, which is nothing but extranuous noise.

**Comment by seven-phases-max** [comment]
>It's not. Please go back and re-read.

With "changing output" I directly mean "child class name extracted as a substring from a parent class name".

Actually I'm a bit surprised that the idea of a technique like this comes form *you* (considering critiques like [this](https://github.com/less/less.js/issues/2436#issuecomment-76174977) and similar).
Because, to be honest, it's nothing but:
"I don't won't to pass an argument to `sqrt` function because it's nothing but extranuous noise, I need a method to extract that value within `sqrt` itself from the callee scope variable of some predefined name".

**Comment by rjgotten** [comment]
> Actually I'm a bit surprised that the idea of a technique like this comes form you (considering critiques like this and similar).
> Because, to be honest, it's nothing but:
> "I don't won't to pass an argument to sqrt function because it's nothing but extranuous noise, I need a method to extract that value within sqrt itself from the callee scope variable of some predefined name".

Detached rulesets in Less query the full call-site scope before the declaration scope when evaluating variables and yes; I still believe that's a stupid thing to do.

Reading back the current selector is *not* the same. 

The critical differences are:
  1. that variable names in your paraphrased quote don't follow a known naming system like CSS class names can follow BEM or similar system with implied hierarchy encoded into the name, and
  2. that such an approach of reading back the selector would only look at the right-hand key-selector (i.e. the deepest level) instead of looking all the way up the chain potentially pulling a value from who knows how far up the scope.

**Comment by rjgotten** [comment]
> Actually, current context including parent selectors is available in functions (this.context.frames should include selectors too unless something wrong with the calling code). 

Then you do have a bug; because `this.context.frames.selectors` seems to always be the 'empty' selector set, i.e. , an array with a single null element ( `[null]` ), when pulled up inside a mixin. :)

**Comment by seven-phases-max** [comment]
>`this.context.frames` should include selectors too unless something wrong with the calling code

Just in case I've just tested [this snippet](https://gist.github.com/seven-phases-max/a927e36da33c73728145) within a basic "node-module" plugin and it works as expected.

**Comment by seven-phases-max** [comment]
>Then you do have a bug; because  this.context.frames.selectors  seems to always be the 'empty' selector set, 

OK, I guess I can see now where the problem is. I referred to `this.context.frames[n].selectors` which in fact contains all the selectors visible in current context (in order of their definition, so it's impossible to actually find which one is the parent of the current scope), while you looked at more proper looking `this.context.selectors` which *does* represent current ruleset selectors (but only inside a ruleset and not anything invoked from there

Hmm... Well, it's complicated. Because `this.context.selectors` seems to be used to represent only the selectors of the ruleset/mixin *definitions* rather than selectors of the scope they are invoked in (so for a mixin defined in the global scope `this.context.selectors` is always empty).
The same actually happens for mixins calls. I.e. when a mixin is called from another mixin `context.selectors` of the call object (`context` argument of `MixinCall.prototype.eval`) is also empty. Though this does not affect anything because any stuff that could refer to parent selectors there (e.g. `&` for a ruleset defined within such mixin) are evaluated after the mixin is expanded.

So technically I suspect this particular field is not what we are looking for (then I was wrong and you can't currently get parent selectors within  function) because if it's supposed to represent definition scope for rulesets and mixins it would be strange to have it to represent callee scope for functions.
And then it has to be some additional entries/code. Btw, and the same most likely will apply to the proposed feature implementation.

**Comment by seven-phases-max** [comment]
>this.context.frames[n].selectors  which in fact contains all the selectors visible in current context (in order of their definition, so it's impossible to actually find which one is the parent of the current scope)

Correcting myself (again, doh! I already posted almost the `parent-selectors` code using  before my previous post but then I was misleaded by a mistake in it),

So actually it seems I was right initiaait's `this.context.frames[n]` is more like what should be used (not sure yet). Sure each frame is actually a context starting from the current one and going upwards to the root (so one of the frames in between those should actually a callee ruleset). I'm not yet sure of what a proper

**Comment by seven-phases-max** [comment]
Sorry, for false closing (I was editing my post, for a third time today, doh!, and accidentaly pushed the wrong button).
Apparently it's `this.context.frames.selectors` to be used and not `this.context.selectors` as the latter represents selectors of the current context *definition* (hense it's empty for mixins defained in the global scope) while the former is the chain of contexts from current scope to global one. Here's almost working [prototype](https://gist.github.com/seven-phases-max/a927e36da33c73728145) - the only probem with it as beside actual parent selector it also include selector of the mixins it's invoked within (if any). Currently I'm trying to find what would be a proper way to filter those.

**Comment by matthew-dean** [comment]
If there's a legitimate bug, yes, by all means let's fix it. My comment was only that there a variety of other supported patterns with no additional complexity that would achieve the same output. But if that wasn't what you were asking about, my mistake.

**Comment by seven-phases-max** [comment]
Btw. I did more tests with the protorype above and so far found the only bug related to detached rulesets. If a DR is defined before its called - everything is fine. However if it's defined after use its `frames` chain seems to be duplicated.

**Comment by rjgotten** [comment]
@seven-phases-max 
Some very interesting stuff here.
I'm going to have to have a look at this over the weekend. Thank you. ^_^
